packages/network-controller/src/provider-api-tests/shared-tests.ts

```diff
@@ -0,0 +1,2673 @@
+      const firstMockParams = [
+        ...new Array(numberOfParameters).fill('some value'),
+      ];
+      const secondMockParams = firstMockParams.slice();
+      secondMockParams[paramIndex] = 'another value';
+      const requests = [
+        {
+          method,
+          params: firstMockParams,
+        },
+        { method, params: secondMockParams },
+      ];
+      const mockResults = [{ blockHash: '0x100' }, { blockHash: '0x200' }];
+
+      await withMockedCommunications({ providerType }, async (comms) => {
+        // The first time a block-cacheable request is made, the latest block
+        // number is retrieved through the block tracker first. It doesn't
+        // matter what this is — it's just used as a cache key.
+        comms.mockNextBlockTrackerRequest();
+        comms.mockRpcCall({
+          request: requests[0],
+          response: { result: mockResults[0] },
+        });
+
+        comms.mockRpcCall({
+          request: requests[1],
+          response: { result: mockResults[1] },
+        });
+
+        const results = await withNetworkClient(
+          { providerType },
+          ({ makeRpcCallsInSeries }) => makeRpcCallsInSeries(requests),
+        );
+
+        expect(results).toStrictEqual([mockResults[0], mockResults[1]]);
+      });
+    });
+  }
+};
+
+type TestsForRpcMethodSupportingBlockParam = {
+  blockParamIndex: number;
+  numberOfParameters: number;
+  providerType: 'infura' | 'custom';
+};
+
+export const testsForRpcMethodSupportingBlockParam = (
+  method: string,
+  {
+    blockParamIndex,
+    numberOfParameters,
+    providerType,
+  }: TestsForRpcMethodSupportingBlockParam,
+) => {
+  describe.each([
+    ['given no block tag', undefined],
+    ['given a block tag of "latest"', 'latest'],
+  ])('%s', (_desc, blockParam) => {
+    it('does not hit the RPC endpoint more than once for identical requests', async () => {
+      const requests = [
+        {
+          method,
+          params: buildMockParams({ blockParamIndex, blockParam }),
+        },
+        { method, params: buildMockParams({ blockParamIndex, blockParam }) },
+      ];
+      const mockResults = ['first result', 'second result'];
+
+      await withMockedCommunications({ providerType }, async (comms) => {
+        // The first time a block-cacheable request is made, the block-cache
+        // middleware will request the latest block number through the block
+        // tracker to determine the cache key. Later, the block-ref
+        // middleware will request the latest block number again to resolve
+        // the value of "latest", but the block number is cached once made,
+        // so we only need to mock the request once.
+        comms.mockNextBlockTrackerRequest({ blockNumber: '0x100' });
+        // The block-ref middleware will make the request as specified
+        // except that the block param is replaced with the latest block
+        // number.
```

Comment:
@BelfordZ You mentioned this in the PR description, but it looks like previously this was "0x100" (or whatever block number the request above returned), and now it's either `null` or `'latest`'. The `block-ref` middleware isn't present in the `web3-provider-engine` stack, so that makes sense.

**Nit:** If we are interested in keeping these tests, then we may want to remove the explicit block number as well as references to the block-ref middleware in the comments, so:

``` suggestion
        // The first time a block-cacheable request is made, the block-cache
        // middleware will request the latest block number through the block
        // tracker to determine the cache key.
        comms.mockNextBlockTrackerRequest();
```

Same goes for every other instance of this.